### ISSUE TITLE
Embedding precision in Vector data type

### ISSUE THREAD

**Issue by wmwsagara** [pgvector/pgvector#423]
**1)** We noted in pgvector, the vector data type is defined as follows (src/vector.h):
```
typedef struct Vector
{
 int32		vl_len_;		/* varlena header (do not touch directly!) */
 int16		dim;			/* number of dimensions */
 int16		unused;
 float		x[FLEXIBLE_ARRAY_MEMBER];
} Vector;
```

That is embeddings are C float type. That is quite low precision!

**2)** OpenAI embeddings
OpenAI sends us embeddings of following precision:
0.0009444306806544363 (19 fractional digits)
-0.0003779408325534785 (19 fractional digits)

C float has 6 decimal digits precision and C double has 15 decimal digits precision.

**3)** We are trying to use pgvector with a large document collection. And we intend to use Hierarchical Navigable Small Worlds (HNSW) for indexing. This significant loss of precision is a concern.

**4)** GPT4 opinion
We checked with GPT4. It's opinion is this level of precision loss when build HNSW index and search may have implication on accuracy. Please see below the section titled "GPT4 opinion".

**5)** RAM
We understand memory requirements of HNSW index is related to the number of Edges Per Node (m). But the quality of HNSW index is surely related to the precision of embeddings. 

**6)** Concerns related to upgrade pgvector
We hope that upgrading float to double on Vector data type is good enough. We like to understand the what are the concerns of this move on pgvector side?

**7)** Patch file
Can we submit a patch file to fix the issue?

Best regards
Sagara

**GPT4 opinion**
Question:
[...] What is the implication of this towards the accuracy of HNSW index search by Cosine distance?

GPT4:
The conversion of high-precision NUMERIC(p, s) data to C float type in pgvector, for the purpose of building an HNSW index, has several implications, particularly for the accuracy of cosine distance-based searches. Here's an analysis of the situation:

**Understanding the Precision**

NUMERIC(p, s): This PostgreSQL data type can store numbers with up to p digits, s of which can be after the decimal point. It's a high-precision, decimal-based format.

C float: A standard C float is typically a 32-bit floating-point data type based on the IEEE 754 standard. It has a precision of about 7 decimal digits.

**Implications of Precision Reduction**

Loss of Precision: When high-precision NUMERIC(p, s) values are converted to C floats, there is a significant loss of precision. This could potentially impact the accuracy of similarity calculations, as small differences in high-dimensional space can be meaningful.

Impact on Cosine Distance: Cosine similarity is sensitive to the direction (angle) of the vectors rather than their magnitude. While the loss of precision might not drastically alter the direction of the vectors, in high-dimensional spaces even slight changes can lead to different results. This is especially true for very similar vectors where small differences can change the ranking of nearest neighbors.

Quantization Noise: The conversion introduces a form of quantization noise. This noise might not be uniform across all dimensions and values, potentially leading to biases in how the HNSW index structures the data.

Practical Impact: The practical impact on your specific use case (text document similarity) would depend on the nature of your embeddings. If the embeddings are robust to small perturbations (as many machine learning embeddings are designed to be), the loss in precision might not severely impact the overall performance of the system. However, if your embeddings are sensitive to small changes, the precision loss could be more critical.

**Mitigating Strategies**

Testing and Evaluation: It's important to test and evaluate the HNSW index with actual data and queries to understand the practical impact. Compare the results of similarity searches with the reduced precision against a baseline to quantify the loss in accuracy.

Alternative Data Types: If precision is a critical concern, consider alternative approaches or data types that can better preserve the original precision of the embeddings.

Custom Solutions: In cases where default behavior of tools like pgvector doesn't meet your precision requirements, you might need to consider more customized solutions or modifications to the existing tools.

In summary, while there is a clear loss of precision when converting NUMERIC(p, s) to C float, the practical impact on cosine distance-based searches for your embeddings depends on the nature of the embeddings and your specific accuracy requirements. Testing with your actual data is crucial to assess the real-world implications.

**Comment by hlinnaka** [comment]
I believe OpenAI embeddings are in fact 32-bit floats. See https://github.com/EliahKagan/embed-encode/blob/main/why.md.

**Comment by wmwsagara** [comment]
@hlinnaka, Thank you for the link shared. It cleared our confusion. The issue was the Python framework we used converts OpenAI 32bit floats to 64bit double. We will directly use OpenAI's Create embeddings API endpoint to get embeddings in 32bit float and save to PostgreSQL as real data type (32bit float). Therefore, no loss in the entire pipeline.

Conclusion:
Pgvector's Embedding precision in Vector data type is not required upgrade to double until OpenAI upgrade to double.
